consensus/tendermint/propose.go

```diff
@@ -0,0 +1,246 @@
+package tendermint
+
+func (t *Tendermint[V, H, A]) handleProposal(p Proposal[V, H, A]) {
+	if p.H < t.state.h {
+		return
+	}
+
+	if p.H > t.state.h {
+		if p.H-t.state.h > maxFutureHeight {
+			return
+		}
+
+		if p.R > maxFutureRound {
+			return
+		}
+
+		t.futureMessagesMu.Lock()
+		defer t.futureMessagesMu.Unlock()
+		t.futureMessages.addProposal(p)
+		return
+	}
+
+	if p.R > t.state.r {
+		if p.R-t.state.r > maxFutureRound {
+			return
+		}
+
+		t.futureMessagesMu.Lock()
+		defer t.futureMessagesMu.Unlock()
+
+		t.futureMessages.addProposal(p)
+
+		t.line55(p.R)
+		return
+	}
+
+	// The code below shouldn't panic because it is expected Proposal is well-formed. However, there need to be a way to
+	// distinguish between nil and zero value. This is expected to be handled by the p2p layer.
+	vID := (*p.Value).Hash()
+	validProposal := t.application.Valid(*p.Value)
+	proposalFromProposer := p.Sender == t.validators.Proposer(p.H, p.R)
+	vr := p.ValidRound
+
+	if validProposal {
+		// Add the proposal to the message set even if the sender is not the proposer,
+		// this is because of slahsing purposes
+		t.messages.addProposal(p)
+	}
+
+	_, prevotesForHR, precommitsForHR := t.messages.allMessages(p.H, p.R)
+
+	if t.line49WhenProposalIsReceived(p, precommitsForHR, vID, validProposal, proposalFromProposer) {
+		return
+	}
+
+	if p.R < t.state.r {
+		// Except line 49 all other upon condition which refer to the proposals expect to be acted upon
+		// when the current round is equal to the proposal's round.
+		return
+	}
+
+	t.line22(vr, proposalFromProposer, validProposal, vID)
+	t.line28WhenProposalIsReceived(p, vr, proposalFromProposer, vID, validProposal)
+	t.line36WhenProposalIsReceived(p, validProposal, proposalFromProposer, prevotesForHR, vID)
```

Comment:
I think we can also have better naming for these lines

Reply:
I disagree, we want to have as close mapping to the lines in the paper as possible. Coming up with another name will only convolude the meaning and make it difficult to debug when the name is not clear enough to indicate which lines from the algorithm are being implemented.

Reply:
Ok I will update the names with part of the condition which ends with rule.